larsim/Utils/TruthMatchUtils.cc

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file larsim/Utils/TruthMatchUtils.cc
+ *
+ * @brief Implementation of the TruthMatchUtils functions
+ *
+ * @author Dom Brailsford (d.brailsford@lancaster.ac.uk)
+ *
+ * $Log: $
+ */
+
+//STL
+#include <utility>
+//ART
+#include "canvas/Utilities/Exception.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+//LARSoft
+#include "larsim/MCCheater/BackTrackerService.h"
+
+#include "TruthMatchUtils.h"
+
+namespace
+{
+
+using namespace TruthMatchUtils;
+IDToEDepositMap::const_iterator MaxEDepElementInMap(const IDToEDepositMap &idToEDepMap)
+{
+    IDToEDepositMap::const_iterator highestContribIt(std::max_element(idToEDepMap.begin(), idToEDepMap.end(),
+        [](const std::pair<G4ID, EDeposit>& a, const std::pair<G4ID, EDeposit>& b)->bool{ return
+        std::nextafter(a.second, std::numeric_limits<EDeposit>::lowest()) <= b.second && std::nextafter(a.second, std::numeric_limits<EDeposit>::max()) >= b.second
+        ? 1 : a.second < b.second; } ));
+
+    if (idToEDepMap.end() == highestContribIt)
+    {
+     throw art::Exception(art::errors::LogicError)
+        << "TruthMatchUtils did not manage to find a max element in the g4 ID to energy deposit map.  The map size is: "
+        << idToEDepMap.size();
+    }
+
+    return highestContribIt;
+}
+
+} // namespace
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleID(const art::Ptr<recob::Hit> &pHit, const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+    if (0 == idToEDepMap.size())
```

Comment:
Recommend:
```.cpp
if (idToEDepMap.empty())
```
instead.  And likewise in various places below.

Reply:
Nice one.
I've updated all container size == 0 checks to use `empty()` instead.